pytorch_lightning/core/lightning.py

```diff
@@ -1170,16 +1170,47 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
 
         Override for your own behavior
 
+        It works with ``untoggle_optimizer`` to make sure param_requires_grad_state is properly reset.
+
         Args:
-            optimizer:
-            optimizer_idx:
+            optimizer: Current optimizer used in training_loop
+            optimizer_idx: Current optimizer idx in training_loop
         """
-        for param in self.parameters():
-            param.requires_grad = False
+        param_requires_grad_state = {}
+        # make sure current optimizer is latest to be iterated over.
+        optimizers = [opt for opt in self.optimizers(use_pl_optimizer=False) if opt != optimizer] + [optimizer]
+        num_optimizers = len(optimizers) - 1
+        for opt_idx, opt in enumerate(optimizers):
+            for group in opt.param_groups:
+                for param in group['params']:
+                    if num_optimizers == opt_idx:
+                        # If a param appears in 2 optimizers, revert `requires_grad` to before toggle.
+                        if param in param_requires_grad_state:
+                            param.requires_grad = param_requires_grad_state[param]
```

Comment:
@tchaton @awaelchli  the last optimizer's parameters are only looked up from the dictionary, but we haven't added them to the dictionary before this point unless they were contained in another optimizer's params.

Reply:
I think it should be something like this instead, which matches what @jain-anshul suggested in #5292 
```
        param_requires_grad_state = {}
        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
            for group in opt.param_groups:
                for param in group['params']:
                    # If a param appears in multiple optimizers, revert `requires_grad` to before toggle
                    if param in param_requires_grad_state:
                        param.requires_grad = param_requires_grad_state[param]
                    else:
                        param_requires_grad_state[param] = param.requires_grad
                        param.requires_grad = False
        for group in optimizer.param_groups:
            for param in group['params']:
                param.requires_grad = param_requires_grad_state[param]
        self._param_requires_grad_state = param_requires_grad_state
 ```